### PR TITLE
SR-1005: Replace POSIX module by Foundation

### PR DESCRIPTION
**Pull request by Bouke**
A partial implementation of [SR-1005][1], replacing POSIX calls by Foundation calls. This PR addresses most file and directory operations by using NSFileManager and NSURL. To prevent a major overhaul of all usages of these calls, I kept all functions intact but changed the implementation. Some things to consider before merging this PR:

* Resolved: segfaults on Linux (see comment below)
* Cocoa Foundation API differs from SwiftFoundation. I've used `#if os(Linux)` for the interim until this is fixed.
* Errors are now mostly Cocoa's default errors; so any code inspecting the error code should be adjusted (1 occurrence found and updated).
* `unlink` and `rmdir` both use `NSFileManager.removeItem(atPath:)`, which will remove files and directories recursively. The old implementations of `unlink` would only remove files, and `rmdir` would only remove directories non-recursive.
* Temporary folders are now created with appending a GUID to guarantee random folder names.
* `mkdir` returns relative folder name if relative name was passed. Use `realpath(mkdir(...))` when old functionality is wanted; opt-in for absolute paths.
* Unused functions (`*time*`, `readlink`, `stat`) were removed.
* Dropped `relativeTo` argument of `symlink` as it was not used; all usages used an absolute path.

Not covered by this PR:

* Replacing `popen` and `system` by NSTask, see #303
* File handling (fopen fputs etc), see #292

Dependencies:
* [x] Fix in NSFileManager.moveItem, see apple/swift-corelibs-foundation#350

[1]: https://bugs.swift.org/browse/SR-1005

### CODE REVIEW

**Comment by ddunbar** [comment]
Awesome!

For test crashes, you should be able to run the tests in a debugger to see what is actually happening.

Can you file bugs for the places where the Foundation API differs across platforms?

It would be good to identify the source of the performance regression, as well.

**Comment by Bouke** [comment]
The segfault as reported by the LLDB debugger, however I'm not sure how to proceed.

    * thread #1: tid = 13474, 0x00007ffff7b4b708 libswiftCore.so`swift_getErrorValue + 8, name = 'swift-build', stop reason = signal SIGSEGV: invalid address (fault address: 0x90)
        frame #0: 0x00007ffff7b4b708 libswiftCore.so`swift_getErrorValue + 8
    libswiftCore.so`swift_getErrorValue:
    ->  0x7ffff7b4b708 <+8>:  movq   0x90(%rcx), %rcx
        0x7ffff7b4b70f <+15>: movzwl %cx, %esi
        0x7ffff7b4b712 <+18>: leaq   0x20(%rdi,%rsi), %rsi

API differences reported in [SR-1328][1].

I suspect that the performance regressions are expected as the old POSIX calls are very lean. The new calls using Foundation probably contain extra error checking; more layers of indirection. 

[1]: https://bugs.swift.org/browse/SR-1328

**Comment by ddunbar** [comment]
> I suspect that the performance regressions are expected as the old POSIX calls are very lean. The new calls using Foundation probably contain extra error checking; more layers of indirection.

This probably isn't the case. While it is true there is a little more indirection, that is unlikely to be able to add up to 6% given the kind of work we are doing... my guess is something else is going on here. Can you try to identify a particular individual test which has regressed?

**Comment by Bouke** [comment]
> This probably isn't the case. While it is true there is a little more indirection, that is unlikely to be able to add up to 6% given the kind of work we are doing... my guess is something else is going on here. Can you try to identify a particular individual test which has regressed?
That was indeed the case. Some NSFileManager functions performed even slightly better in a quick performance test. The full test suite is now run as fast as before.

**Comment by mxcl** [comment]
Great stuff. Let's deconflict and merge.

**Comment by Bouke** [comment]
Indeed it does conflict, it was written against the latest development toolchain snapshot. The Foundation API is being imported now with new rules, so I'm unable to compile until I have an updated toolchain. I've yet to learn to build my own toolchain, maybe I'll just wait for a new toolchain snapshot.

**Comment by Bouke** [comment]
I've located the segfaults on Linux to be related to `NSFileManager.rename(atPath:toPath:)`, for which I've filed a [bug report][1]. I've updated this PR to have smaller atomic commits, which allowed for `git bisect` to find the segfault. The PR is rebased on master against the updated Foundation API. The PR has some overlap with #292, but I'll resolve that as soon as either one of these PRs are merged.

[1]: https://bugs.swift.org/browse/SR-1380

**Comment by Bouke** [comment]
I've rebased the PR and verified against OSX and Ubuntu 15.10.

**Comment by Bouke** [comment]
I've also moved `rename` over to NSFileManager, making this PR dependent on the merger of apple/swift-corelibs-foundation#350.

**Comment by ddunbar** [comment]
Can you rebase please?

**Comment by ddunbar** [comment]
@swift-ci please test and merge

File: Sources/Utility/unlink.swift [hunk 1]
```diff
@@ -8,12 +8,8 @@
  See http://swift.org/CONTRIBUTORS.txt for Swift project authors
 */
 
-
-import func libc.unlink
-import var libc.errno
+import Foundation
 
 public func unlink(_ path: String) throws {
```
Comment (ddunbar) [inline Sources/Utility/unlink.swift]: This function should be renamed, using `unlink` as a synonym for removing a tree is confusing, as that is not the POSIX behavior.

**Comment by ddunbar** [comment]
@swift-ci please test

**Comment by Bouke** [comment]
Any ideas on why the build failed on OSX? The logs include the following test case, but I'm not sure how that's related?

    ERROR: test_lldbmi_break_insert_function_pending (TestMiBreak.MiBreakTestCase)
       Test that 'lldb-mi --interpreter' works for pending function breakpoints.

**Comment by ddunbar** [comment]
That is unrelated, it sounds like from mail on swift-dev.

**Comment by Bouke** [comment]
Can you retest?

**Comment by ddunbar** [comment]
@swift-ci please test

**Comment by Bouke** [comment]
The failing build on Linux doesn't appear to be related. I've tested on both OSX and Ubuntu 15.10.

**Comment by aciidb0mb3r** [comment]
@Bouke conflicts

**Comment by Bouke** [comment]
@ddunbar @aciidb0mb3r rebased and tested against OSX and Ubuntu 15.10.

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Comment by ddunbar** [comment]
I am going to start pulling over individual parts of this piecemeal for things that make more sense to be implemented on top of Foundation than POSIX wrappers...

**Comment by ddunbar** [comment]
I posted a new PR #355 which cherry picks (and reworks) a couple of commits from this PR.

The ones I took:
 * 07f50bfdc5a1cfa2da434e5643d5e38b6d826b6b [Foundation] Implemented mkdir
 * f5728c69b7369aa7aa80efeeecd70b5a068bbff3 [Foundation] Implemented unlink / rmdir / rmtree

The ones I abandoned, and why:

 * 3c673df3239c7ebefb391cefc228a81d55717a1a [Foundation] Implemented POSIX.realpath

    Foundation implementation is more complicated than POSIX.

 * 24f287861beeb927ace4788d4c789887dd35fb0d [Foundation] Implemented chdir

    Foundation implementation isn't better than POSIX one, and we need to change interface slightly from what it is in either case:

 * e1e785286fe5f63097b251c72637ea696099c927 [Foundation] Implemented getcwd

    Foundation implementation doesn't handle all cases we do (missing cwd).

* b5a4a85043a71d71dd9dbc70a7b63c161ca03116 [Foundation] Implemented symlink,  c8bea507adb6cc0a8cdd51f0ae8a71c52816c47b [Foundation] Implemented rename

    We may eventually have a need for using the `...at` POSIX facilities directly (which Foundation doesn't expose), so I'd rather keep the implementation which does that in our own code.

* c56dca2bf367b11eab6d6b204543d42f0cfc3a86 [Foundation] Implemented mkdtemp

    We need to change the current implementation, but I prefer to continue to use mkdtemp at least versus reimplementing it, so I will rework this part later.

**Pull request closed without merging**